src/jdk.incubator.code/share/classes/jdk/incubator/code/dialect/core/CoreOp.java

```diff
@@ -318,6 +320,84 @@ public Block.Builder lower(Block.Builder b, CodeTransformer _ignore) {
             b.rebind(b.context(), CodeTransformer.LOWERING_TRANSFORMER).op(this);
             return b;
         }
+
+        static CoreOp.FuncOp invokeToFuncOp(JavaOp.InvokeOp invokeOp, MethodHandles.Lookup l) {
+            try {
+                Method method = invokeOp.invokeDescriptor().resolveToMethod(l);
+                return Op.ofMethod(method).orElse(null);
+            } catch (ReflectiveOperationException e) {
+                throw new IllegalStateException("Could not resolve invokeOp to method");
+            }
+        }
+
+        /**
+         * Returns a moduleOp with the given funcOp as the root.
+         * The funcOps in the moduleOp functionTable are returned in the order encountered within the funcOp.
+         */
+        public static CoreOp.ModuleOp ofFuncOp(CoreOp.FuncOp f, MethodHandles.Lookup l) {
+            Map<String, FuncOp> visited = new HashMap<>();
+            List<CoreOp.FuncOp> funcs = new ArrayList<>();
+            Deque<CoreOp.FuncOp> stack = new LinkedList<>();
+
+            int idx = 0;
+            stack.push(f);
+            while (!stack.isEmpty()) {
+                CoreOp.FuncOp cur = stack.pop();
+
+                String name = cur.funcName() + "::" + cur.invokableType();
+                if (visited.containsKey(name)) {
```

Comment:
Make `visited` a `Map<FuncOp, String>` which stores the name used for the transformed function in the module op and the name used for the func call op. No need to throw as you will always create a unique name based on the counter.